### ISSUE TITLE
Support casting of `bytes` to value types

### ISSUE THREAD

**Issue by lukehutch** [ethereum/solidity#13211]
## Abstract

Casting from `bytes` to a value type should unpack just the required number of bytes from the bytes array. For example, given `bytes memory b`, it should be possible to do `bytes4 b4 = bytes4(b);`

## Motivation

The primary motivation for this is for decoding the return value of low-level calls, when there is only one return value:

```solidity
    (bool success, bytes memory returnData) = addr.call(abiEncoding);
```

## Specification

Given `returnData` as returned from a `call` as shown above, there needs to be a way to easily read a single value of a given type from the return data, without using assembly.

The assembly version would be, e.g. for `bytes4` return type,

```solidity
    if (success) {
        bytes4 val;
        if (returnData.length >= 4) {
            assembly {
                val := mload(add(returnData, 32))
            }
        } else {
            val = 0;
        }
        // do something with val here
    }
```

This sort of pattern seems to be needed for any return type other than `string`, because (as far as I can tell) you can't accomplish this by merely casting, e.g. if you expect a `bytes4` as the return value, you can't just go `bytes4(returnData)`.

However assembly sections are always flagged in security audits as potential issues, and additionally, assembly potentially makes code unportable in future to other blockchains or other VMs, etc.

Casting should work for any value type, e.g. if a function call returns a boolean, you should be able to do `bool flag = bool(returnData);`. Currently this requires the same pattern shown above, with `bytes4` replaced with `bool`, or if implementing this in pure Solidity, something like:

```solidity
    bool retVal = success && returnData.length > 0 ? returnData[0] != 0 : false;
```

it would be much better to be able to do something like 

```solidity
    bool retVal = success && bool(returnData);
```

Of course if `returnData.length` is too short to hold the target type, then the cast should throw.

## Backwards Compatibility

This would be fully backwards-compatible, as currently the type of casting suggested above raises an error.

**Comment by hrkrshnn** [comment]
> For example, given bytes memory b, it should be possible to do bytes4 b4 = bytes4(b);

This is already possible since 0.8.5, therefore closing.

**Comment by lukehutch** [comment]
@hrkrshnn this definitely does not work, even on Solidity 0.8.15.  `bool(bytesVal)` gives `TypeError: Explicit type conversion not allowed from "bytes memory" to "bool".`  `bytes4(bytesVal)` gives `TypeError: Type bytes memory is not implicitly convertible to expected type bytes4.`

**Comment by hrkrshnn** [comment]
@lukehutch Are you sure? Here's a remix link that shows the conversion working: https://remix.ethereum.org/#gist=c1177b487cb41b7dfda95553a022a728

From your error message, you seem to be "implicitly" converting, something like `bytes4 x = b` would be disallowed. But `bytes4 x = bytes4(b)` would be allowed.

Regarding `bool`, that seems independent of the issue. Do you really want conversions to `bool`?

**Comment by lukehutch** [comment]
@hrkrshnn Yes, I'm sure (your `bytes4` example works, but changing the type to `bool` definitely doesn't) -- the following in your remix gist gives an error:

```solidity
contract C {
    function convert(bytes memory m) external pure returns (bool) {
        return bool(m);
    }
}
```

```
TypeError: Explicit type conversion not allowed from "bytes memory" to "bool".
--> gist-c1177b487cb41b7dfda95553a022a728/convert.sol:4:16:
|
4 | return bool(m);
| ^^^^^^^
```

Other types such as `uint256` also cannot be loaded from the beginning of a `bytes` array in the same way, which leads to the need to drop assembly code into a program just to parse out typed data from `bytes`.

I had a contract where this conversion wasn't working for `bytes4` either, but I'll have to roll back to an earlier version and try to cut it down to a minimal reproducer.

**Comment by hrkrshnn** [comment]
I would say a bytes to bool conversion is rather ambiguous, and already disallowed by the compiler. How do you expect the conversion to work?

**Comment by lukehutch** [comment]
I would expect a bool conversion to work exactly the same as the pattern I gave earlier:

```solidity
    if (success) {
        bool val;
        if (returnData.length >= 1) {
            assembly {
                val := mload(add(returnData, 32))
            }
        } else {
            val = false;
        }
        // do something with val here
    }
```

I assume this loads only the first byte of the `returnData` array, and if it is nonzero, the resulting value is true.

Is the ambiguity you raise referring to whether the bool is packed or not? (I assume the packed version is stored as one byte, and the non-packed encoding stores it as a `uint256`?) 

Maybe I'm going crazy but I tried to reproduce this with `bytes4`, and couldn't reproduce the problem, even though I first saw this conversion limitation with `bytes4`, and ran into it several times. I just can't seem to trigger it anymore.

**Comment by lukehutch** [comment]
This needs to be supported for `uint256` at the very least, but probably for all integer types.

In the more general case, it would be great to be able to do something like:

`(uint256 x, uint32 y, bool z) = (uint256, uint32, bool)(myBytesArray);`

**Comment by hrkrshnn** [comment]
You can already achieve this by using `abi.decode(bytesVal, (bool))`. Similarly for general tuple types.

**Comment by cameel** [comment]
Yeah, just like @hrkrshnn said, `abi.decode()` is the general solution here. In a general case you need actual ABI decoding because not all types have the layout on the stack that's identical to how they're encoded in calldata/returndata.

But also, overall this looks to me like you're trying to use a low-level `call()` but still want it to work like a high-level call that decodes returned data into Solidity types. Why are you not doing a high-level call instead? Is it because of extra checks that add gas overhead? I.e. are you actually after something like #11125?

**Comment by lukehutch** [comment]
Right, I have been using low-level calls, primarily because I wanted the result of the call in the `(bool success, bytes memory returnData)` return format, so that the whole transaction wouldn't be reverted if a call failed, and so that if it did fail, I could decode exactly why it failed and react accordingly. Yes, #11125 is closely related to the concerns I had that led me to calling external contracts in this way.

**Comment by cameel** [comment]
Ah, if it's about handling all possible errors then it sounds more like #12725. We really need to address that one. It would help if you could share a bit about your particular use case in that issue. It's currently stuck in the design phase because to make it super granular we'll need to bend the syntax in ways that are not great, but we also don't know how much granularity is actually necessary in practice and maybe we can get away with less.

**Comment by lukehutch** [comment]
> It would help if you could share a bit about your particular use case in that issue.

I was trying to call a range of external contracts implementing interfaces that all had different rules about what should be done if the callee address was an EOA, and/or if the callee function was not implemented in a contract, and/or if the callee function reverted, and/or if the contract function did not revert but returned the wrong value.

The `catch Error(string memory reason)` and `catch (bytes memory data)` constructs in Solidity were not very helpful, because `(bytes memory data)` still had to be decoded anyway, and some call situations (e.g. trying to call a function on an EOA) reverted with empty data, rather than with `Error`.

In the end it got complex enough trying to implement the required logic every time that I ended up creating a single all-singing external caller function that just worked directly with

```solidity
(bool success, bytes memory returnData) = addr.call{value: ethValue}(abiEncoding);
```

then implemented all the complex return logic that I needed based on the two return values.

I ended up reverting this back to `ISomeInterface(addr).someFunction()` calls because security auditors didn't like the use of `address.call`, and I just had to add some checks before each call, such as `require(isContract(addr), "Not a contract")` to prevent cases where external calls would revert the transaction without any reason code. But I liked the robustness of the way I had it implemented before.
